data/transactions/logic/assembler_test.go

```diff
@@ -3159,3 +3176,158 @@ warning 2
 	expected = "1 error: 42: super annoying error\n"
 	assertWithMsg(t, expected, b)
 }
+
+func TestLineTokens(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	ops := newOpStream(8)
+	err := ops.assemble(`#pragma version 8
+// put 1 on the stack
+int 1// this is the one 
+return
+`)
+	if err != nil {
+		t.Logf("Errors: %+v", ops.Errors)
+		t.Fail()
+	}
+
+	expectedLines := []lineTokens{
+		{
+			{kind: tokenData, str: "#pragma", col: 0, line: 1},
+			{kind: tokenData, str: "version", col: 8, line: 1},
+			{kind: tokenData, str: "8", col: 16, line: 1},
+		},
+		{
+			{kind: tokenComment, str: "put 1 on the stack", col: 0, line: 2},
+		},
+		{
+			{kind: tokenData, str: "int", col: 0, line: 3},
+			{kind: tokenData, str: "1", col: 4, line: 3},
+			{kind: tokenComment, str: "this is the one", col: 5, line: 3},
+		},
+		{
+			{kind: tokenData, str: "return", col: 0, line: 4},
+		},
+	}
+
+	assert.Equal(t, expectedLines, ops.Lines)
+
+	expectedFilteredLines := []lineTokens{
+		{
+			{kind: tokenData, str: "#pragma", col: 0, line: 1},
+			{kind: tokenData, str: "version", col: 8, line: 1},
+			{kind: tokenData, str: "8", col: 16, line: 1},
+		},
+		nil,
+		{
+			{kind: tokenData, str: "int", col: 0, line: 3},
+			{kind: tokenData, str: "1", col: 4, line: 3},
+		},
+		{
+			{kind: tokenData, str: "return", col: 0, line: 4},
+		},
+	}
+	var filtered []lineTokens
+	for _, lts := range ops.Lines {
+		filtered = append(filtered, lts.filter(tokenComment))
+	}
+	assert.Equal(t, expectedFilteredLines, filtered)
+
+}
+
+func TestStackSnapshot(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	ops, err := assembleWithTrace(`#pragma version 8
+int 1
+int 1
+byte 0xdead
+pop
+pop
+return`, 8)
+
+	if err != nil {
+		t.Logf("Errors: %+v", ops.Errors)
+		t.Fail()
+	}
+
+	// for idx, lts := range ops.Lines {
+	// 	t.Logf("| %-20s | %-20v |", lts, ops.stacks[idx])
+	// }
+
+	expected := []StackTypes{
+		{},
+		{StackUint64},
+		{StackUint64, StackUint64},
+		{StackUint64, StackUint64, StackBytes.narrowed(static(2))},
+		{StackUint64, StackUint64},
+		{StackUint64},
+		{},
+	}
+
+	assert.Equal(t, expected, ops.stacks)
+}
+
+func TestStackSnapshotCallsub(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+	src := `#pragma version 8
+ int 1
+ int 1
+ callsub lbl
+ int 5
+ ==
+ byte 0xbeef
+ swap
+ assert
+ pop
+ return
+
+ lbl:
+ 	proto 0 1
+ 	int 2
+ 	int 3
+ 	+
+ 	retsub
+ `
+
+	ops, err := assembleWithTrace(src, 8)
+
+	if err != nil {
+		t.Logf("Errors: %+v", ops.Errors)
+		t.Fail()
+	}
+
+	for idx, lts := range ops.Lines {
+		t.Logf("| %-20s | %-20v |", lts, ops.stacks[idx].Reverse())
+	}
+
+	t.Logf("%d %+v", len(ops.stacks), ops.stacks)
+	t.Logf("%d %+v", len(ops.Lines), ops.Lines)
```

Comment:
These logs should be removed but wanted to make it easy to see how it might look if some client program wanted to use them